sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/ProactiveOpenConnectionsProcessor.java

```diff
@@ -0,0 +1,315 @@
+            OpenConnectionTask op,
+            ShouldRetryResult retryResult) {
+        if (retryResult.backOffTime == Duration.ZERO || retryResult.backOffTime == null) {
+            this.submitOpenConnectionWithinLoopInternal(op);
+            return Mono.empty();
+        } else {
+            return Mono
+                    .delay(retryResult.backOffTime)
+                    .flatMap(ignore -> {
+                        this.submitOpenConnectionWithinLoopInternal(op);
+                        return Mono.empty();
+                    });
+        }
+    }
+
+    // when the flux associated with openConnectionsTaskSink is cancelled
+    // this method provides for a way to resume emitting pushed tasks to
+    // the sink
+    private void instantiateOpenConnectionsPublisher() {
+        logger.debug("Re-instantiate open connections task sink");
+
+        if (this.isConcurrencySwitchFlow.get()) {
+            openConnectionsTaskSink = openConnectionsTaskSinkBackUp;
+            openConnectionsTaskSinkBackUp = Sinks.many().multicast().onBackpressureBuffer(OPEN_CONNECTION_SINK_BUFFER_SIZE);
+        } else {
+            openConnectionsTaskSink = Sinks.many().multicast().onBackpressureBuffer(OPEN_CONNECTION_SINK_BUFFER_SIZE);
+        }
+    }
+
+    private void forceDefensiveOpenConnections() {
+        if (aggressivenessHint.get() == WarmupFlowAggressivenessHint.AGGRESSIVE) {
+            aggressivenessHint.set(WarmupFlowAggressivenessHint.DEFENSIVE);
+        }
+    }
+
+    private void completeSink(Sinks.Many<OpenConnectionTask> sink) {
+        Sinks.EmitResult completeEmitResult = sink.tryEmitComplete();
+
+        if (completeEmitResult == Sinks.EmitResult.OK) {
+            logger.debug("Sink completed.");
+        } else if (completeEmitResult == Sinks.EmitResult.FAIL_CANCELLED ||
+                completeEmitResult == Sinks.EmitResult.FAIL_TERMINATED) {
+            logger.debug("Sink already completed, EmitResult: {}", completeEmitResult);
+        } else {
+            logger.warn("Sink completion failed, EmitResult: {}", completeEmitResult);
+        }
+    }
+
+    private static class ConcurrencyConfiguration {
+        final int openConnectionOperationEmissionConcurrency;
+        final int openConnectionExecutionConcurrency;
+
+        public ConcurrencyConfiguration(int openConnectionOperationEmissionConcurrency, int openConnectionExecutionConcurrency) {
+            this.openConnectionOperationEmissionConcurrency = openConnectionOperationEmissionConcurrency;
+            this.openConnectionExecutionConcurrency = openConnectionExecutionConcurrency;
+        }
+    }
+
+    private static class SerializedEmitFailureHandler implements Sinks.EmitFailureHandler {
+
+        @Override
+        public boolean onEmitFailure(SignalType signalType, Sinks.EmitResult emitResult) {
+            if (emitResult.equals(Sinks.EmitResult.FAIL_NON_SERIALIZED) || emitResult.equals(Sinks.EmitResult.FAIL_OVERFLOW)) {
```

Comment:
what about the FAIL_NO_SUBSCRIBER error

Reply:
FAIL_NO_SUBSCRIBER simply gets ignored / dropped.